### PR TITLE
[feat] FlashInfer CuteDSL MegaMoE integration 

### PR DESCRIPTION
**Pull request by jdebache**
## Purpose

This changeset attempts to use the recently added CuteDSL MegaMoE kernels from FlashInfer to accelerate large batch size (prefill mostly) inference. We will evaluate on several models to exercise as many paths as possible.

It is effectively a generalization of https://github.com/vllm-project/vllm/pull/49636 beyond DeepSeekV4.

## Test Plan

To test this feature, we evaluate accuracy against baseline with the GSM8K eval and we measured performance using `vllm bench serve` in simulated prefill scenarios (setting OSL to 1), where these kernels should excel. We chose a variety of models to attempt to cover as many code paths as possible.

All runs were performed on 4xGB200 whenever possible. For DeepSeekV4 Pro, we instead ran on 8xB200.

The servers were started with variations of the following, adjusted per model:

**Baseline server**

```bash
vllm serve \
  $MODEL_PATH \
  --served-model-name evaluation-model \
  --tensor-parallel-size 1 --data-parallel-size 4 \
  --enable-expert-parallel --language-model-only --enforce-eager \
  --moe-backend flashinfer_cutedsl \
  --max-model-len 8192 --max-num-seqs 128 \
  --max-num-batched-tokens 16384 --gpu-memory-utilization 0.9 \
  --no-enable-prefix-caching --no-enable-flashinfer-autotune \
  --seed 42 --port 8000
```

**Candidate server**

```bash
vllm serve \
  $MODEL_PATH \
  --served-model-name evaluation-model \
  --tensor-parallel-size 1 --data-parallel-size 4 \
  --enable-expert-parallel --language-model-only --enforce-eager \
  --moe-backend flashinfer_moe_ep_cutedsl \
  --max-model-len 8192 --max-num-seqs 128 \
  --max-num-batched-tokens 16384 --gpu-memory-utilization 0.9 \
  --no-enable-prefix-caching --no-enable-flashinfer-autotune \
  --seed 42 --port 8000
```

Benchmarking load was generated with:

```bash
vllm bench serve \
  --backend openai --host 127.0.0.1 --port 8000 \
  --endpoint /v1/completions --model evaluation-model \
  --tokenizer $MODEL_PATH \
  --dataset-name random --random-input-len 4096 \
  --random-output-len 1 --random-range-ratio 0 \
  --num-prompts 1024 --num-warmups 256 \
  --max-concurrency 128 --request-rate inf --ignore-eos \
  --temperature 0 --seed 42 \
  --percentile-metrics ttft,e2el --metric-percentiles 50,90,99 \
  --save-result --save-detailed \
  --result-dir $OUTPUT_PATH\
  --result-filename prefill-results.json
```

GSM8K was run using the vLLM provided script:

```bash
python \
  tests/evals/gsm8k/gsm8k_eval.py \
  --host http://127.0.0.1 --port 8000 \
  --num-questions 1319 --num-shots 5 --max-tokens 256 \
  --temperature 0 --seed 42 --max-concurrency 128 \
  --save-results $OUTPUT_PATH/gsm8k_out.json
```

#### GSM8K results

| Metric | Baseline | Candidate | Delta |
|---|---:|---:|---:|
| Accuracy | 92.4185% | 92.4185% | 0.0000 pp |
| Correct | 1,219 / 1,319 | 1,219 / 1,319 | 0 |
| Invalid responses | 1 / 1,319 | 0 / 1,319 | -1 |

#### Benchmarking results

| Metric | Baseline | Candidate | Candidate delta |
|---|---:|---:|---:|
| Request throughput | 13.4555 req/s | 15.1372 req/s | +12.50% |
| Mean TTFT | 8,978.92 ms | 7,943.40 ms | -11.53% |
| Median TTFT | 9,437.29 ms | 8,378.39 ms | -11.22% |
| P90 TTFT | 10,580.61 ms | 9,413.25 ms | -11.03% |
| P99 TTFT | 10,839.94 ms | 9,535.96 ms | -12.03% |

#### GSM8K results

| Metric | Baseline | Candidate | Delta |
|---|---:|---:|---:|
| Accuracy | 89.6892% | 90.3715% | +0.6823 pp |
| Correct | 1,183 / 1,319 | 1,192 / 1,319 | +9 |
| Invalid responses | 0 / 1,319 | 0 / 1,319 | 0 |

#### Benchmarking results

| Metric | Baseline | Candidate | Candidate delta |
|---|---:|---:|---:|
| Request throughput | 65.5500 req/s | 86.5584 req/s | +32.05% |
| Mean TTFT | 919.42 ms | 696.35 ms | -24.26% |
| Median TTFT | 943.83 ms | 712.63 ms | -24.50% |
| P90 TTFT | 1,092.19 ms | 813.35 ms | -25.53% |
| P99 TTFT | 1,181.90 ms | 904.35 ms | -23.48% |

#### GSM8K results

| Metric | Baseline | Candidate | Delta |
|---|---:|---:|---:|
| Accuracy | 93.6315% | 94.2381% | +0.6065 pp |
| Correct | 1,235 / 1,319 | 1,243 / 1,319 | +8 |
| Invalid responses | 0 / 1,319 | 0 / 1,319 | 0 |

#### Benchmarking results

| Metric | Baseline | Candidate | Candidate delta |
|---|---:|---:|---:|
| Request throughput | 11.2963 req/s | 12.0641 req/s | +6.80% |
| Mean TTFT | 5,356.16 ms | 5,031.09 ms | -6.07% |
| Median TTFT | 5,574.72 ms | 5,224.03 ms | -6.29% |
| P90 TTFT | 6,188.37 ms | 5,676.57 ms | -8.27% |
| P99 TTFT | 6,977.22 ms | 6,563.66 ms | -5.93% |

### DeepSeek-V4-Flash vs previous MegaMoE integration

Note that for this one, no change is expected, since the baseline is already using MegaMoE, but based on the previous implementation from https://github.com/vllm-project/vllm/pull/49636 beyond DeepSeekV4.

#### GSM8K results

| Metric | Baseline | Candidate | Delta |
|---|---:|---:|---:|
| Accuracy | 94.0864% | 94.3139% | +0.2274 pp |
| Correct | 1,241 / 1,319 | 1,244 / 1,319 | +3 |
| Invalid responses | 0 / 1,319 | 0 / 1,319 | 0 |

#### Benchmarking results

| Metric | Baseline | Candidate | Candidate delta |
|---|---:|---:|---:|
| Request throughput | 33.1389 req/s | 33.0759 req/s | -0.19% |
| Mean TTFT | 1,827.14 ms | 1,828.52 ms | +0.08% |
| Median TTFT | 1,878.42 ms | 1,878.35 ms | 0.00% |
| P90 TTFT | 2,145.42 ms | 2,154.42 ms | +0.42% |
| P99 TTFT | 2,370.58 ms | 2,353.44 ms | -0.72% |

### DeepSeek-V4-Pro (TP1 / DP8 / EP8) vs previous MegaMoE integration

Note: the improvement here is a bit unexpected, but in the right direction. This could be noise caused by "lucky" vs "unlucky" kv-cache thrashing, in turn due to large model + large symmetric memory allocation from MegaMoE.

#### GSM8K results

| Metric | Baseline | Candidate | Delta |
|---|---:|---:|---:|
| Accuracy | 94.1622% | 94.0864% | -0.0758 pp |
| Correct | 1,242 / 1,319 | 1,241 / 1,319 | -1 |
| Invalid responses | 0 / 1,319 | 0 / 1,319 | 0 |

#### Benchmarking results (ISL 8192 +-25%, OSL 1, concurrency 24, 5 reps x 256 requests)

| Metric | Baseline | Candidate | Candidate delta |
|---|---:|---:|---:|
| Request throughput | 6.3401 req/s | 6.8855 req/s | +8.60% |
| Mean TTFT | 3,600.24 ms | 3,343.65 ms | -7.13% |
| Median TTFT | 3,365.89 ms | 3,120.92 ms | -7.28% |
| P90 TTFT | 5,876.33 ms | 4,685.86 ms | -20.26% |
| P99 TTFT | 8,205.04 ms | 6,200.03 ms | -24.44% |

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -6649,12 +6649,28 @@ def shutdown(self) -> None:
...
+        )
+        if clear_graphs_before_distributed_teardown:
+            # MegaMoE communicators own pointers captured by CUDA graphs. Drop
+            # those graphs before collective communicator teardown. ROCm also
+            # requires eager graph destruction to avoid delayed HSA faults.
             CUDAGraphWrapper.clear_all_graphs()
             BreakableCUDAGraphWrapper.clear_all_graphs()
+            if has_flashinfer_megamoe and self.encoder_cudagraph_manager is not None:
+                self.encoder_cudagraph_manager.clear()
             self.encoder_cudagraph_manager = None
```
[the remark this answers is not in the document]
Comment (jdebache) [reply vllm/v1/worker/gpu_model_runner.py]: Removed this logic, apparently not needed.

File: vllm/model_executor/layers/fused_moe/flashinfer_moe_ep.py [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+        w13=w13,
+        w2=w2,
+        w13_scale=folded_w13_scale,
+        w2_scale=w2_scale,
+    )
+    epilogue = FlashInferMoeEpEpilogue(
+        fc1_alpha=gate_scale.clone().contiguous(),
+        fc2_alpha=w2_scale_2.float().clone().contiguous(),
+    )
+    return weights, epilogue
```
[the remark this answers is not in the document]
Comment (jdebache) [reply vllm/model_executor/layers/fused_moe/flashinfer_moe_ep.py:224]: MegaMoE quantizes activations dynamically (input_norm_const and fc1_norm_const stay 1.0), so ModelOpt's static input scales are unused and the alphas are the weight_scale_2 dequant factors alone.

File: vllm/config/kernel.py [hunk 3]
```diff
@@ -147,21 +147,19 @@ def with_default(
...
     }
 )
 
-# Backends that run a mega-MoE model path (fused expert module plus
-# prepare_megamoe routing): vLLM's native deep_gemm path, which any model
-# with a mega-MoE module may use (DeepSeek-V4, Kimi K3), plus the flashinfer
-# moe_ep variants.
-MEGA_MOE_BACKENDS = frozenset({"deep_gemm_mega_moe"}) | FLASHINFER_MOE_EP_BACKENDS
+# Backends that use a model-specific mega-MoE module and routing path.
+MEGA_MOE_BACKENDS = frozenset({"deep_gemm_mega_moe"})
```
Comment (tlrmchlsmth) [inline vllm/config/kernel.py:155]: Why split these out? It makes the name `MEGA_MOE_BACKENDS` confusing. Also means that we need to handle both of them in cases like this

```
    moe_needs_token_sharded_input = (
        moe_backend in MEGA_MOE_BACKENDS
        or moe_backend in FLASHINFER_MOE_EP_BACKENDS
        or parallel_config.data_parallel_size > 1
    )
```

File: vllm/config/kernel.py [hunk 4]
```diff
@@ -187,8 +185,8 @@ def with_default(
 def validate_flashinfer_moe_ep_model(
     moe_backend: str, architectures: Iterable[str]
 ) -> None:
-    """Reject flashinfer moe_ep backends for models that lack the FI path."""
-    if moe_backend not in FLASHINFER_MOE_EP_BACKENDS:
+    """Reject architecture-specific FlashInfer MoE-EP backends."""
```
Comment (tlrmchlsmth) [inline vllm/config/kernel.py]: nit
```suggestion
    """Reject model-specific FlashInfer MoE-EP backends."""
```

since architecture can easily mean hardware architecture

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 5]
```diff
@@ -1072,7 +1148,7 @@ def get_fused_moe_quant_config(self, layer: RoutedExperts) -> FusedMoEQuantConfi
 
     @property
     def supports_eplb(self) -> bool:
-        return True
+        return not self.use_flashinfer_moe_ep
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/quantization/modelopt.py]: Did you run into any issues with EPLB support?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w4a4_nvfp4.py [hunk 6]
```diff
@@ -44,6 +54,30 @@
...
+
+
+def _global_exact_scale(local_scales: torch.Tensor) -> float:
+    local_min, local_max = torch.aminmax(local_scales)
+    bounds = torch.stack((local_max, -local_min))
+    torch.distributed.all_reduce(
+        bounds,
+        op=torch.distributed.ReduceOp.MAX,
+        group=get_ep_group().device_group,
+    )
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_w4a4_nvfp4.py:71]: I don't love the allreduce during weight loading since it could cause issues, although I see that there is some precedent for it. 

Can we get around it? And @mgoin do you have any thoughts?

File: vllm/config/kernel.py (hunk 3 above)
Comment (jdebache) [reply vllm/config/kernel.py:155]: This one is a bit trickier than it looks. On `main`, the MegaMoE implementation ('native' DeepGEMM in vLLM, FlashInfer's DeepGEMM or FlashInfer CuteDSL) doesn't really matter: you always have the modelling code creating its own stuff instead of delegating to the fused_moe + quantization path.

On this branch, the FlashInfer implementations go through fused_moe, but the 'native' DeepGEMM one still does its own thing. That's why we need to discriminate.

I think the correct fix is to make 'native' DeepGEMM go through fused_moe too, but that might make this PR even bigger than it already is. Could be a couple thousands lines more to touch.

Alternatively, I can try to make the names a bit clearer? Still having `MEGA_MOE_BACKENDS` refer to all the mega moe backends, while having something like `NATIVE_MOE_BACKENDS` for the ones that are not used through fused_moe.

@tlrmchlsmth what do you think?

Comment (tlrmchlsmth) [reply vllm/config/kernel.py:155]: >I think the correct fix is to make 'native' DeepGEMM go through fused_moe too, but that might make this PR even bigger than it already is. Could be a couple thousands lines more to touch.

I agree with this

Comment (tlrmchlsmth) [reply vllm/config/kernel.py:155]: >Alternatively, I can try to make the names a bit clearer? Still having MEGA_MOE_BACKENDS refer to all the mega moe backends, while having something like NATIVE_MOE_BACKENDS for the ones that are not used through fused_moe.

Agree with this one too!

File: vllm/model_executor/layers/quantization/modelopt.py (hunk 5 above)
Comment (jdebache) [reply vllm/model_executor/layers/quantization/modelopt.py]: Tried it out, it worked with minimal changes.

**Comment by jdebache** [comment]
@tlrmchlsmth : I've renamed the mega-moe related constants / properties and tested / enabled EPLB.

The only open thread is https://github.com/vllm-project/vllm/pull/54049/changes#r4032072990. My opinion there is that the framework should respect what the checkpoint provides and not need to do these reductions. I.e. if the checkpoint provides one scale per expert, the framework should not take the max. But I don't think this change will fit in this PR.

Once the changes get approved, I would re-run the evals just to make sure.

File: vllm/model_executor/layers/fused_moe/direct_kernel.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+from typing import Protocol
+
+import torch
+
+
+class FusedMoEDirectKernel(Protocol):
```
Comment (bnellnm) [inline vllm/model_executor/layers/fused_moe/direct_kernel.py]: I think we should avoid adding a totally separate kernel class into the MoE layer.  All the existing machinery expects `FusedMoEKernels` and having to add exceptions for a new kernel type will make things harder to maintain.  

We could either modify the `FusedMoEKernel` hierarchy to allow direct kernels or do something like in https://github.com/vllm-project/vllm/pull/40833 where DeepGemm MegaMoE is implemented as a modular kernel that only works with the no-op prepare/finalize classes.

**Comment by jdebache** [comment]
@bnellnm I followed your advice and went with the same approach as in #40833.

While testing DSV4.1 and dspark (since I'm also touching these files), I found a couple of issues which I fixed in the latest commit 2a84a89972a6654791efcab4b4d763298ffeabe0.

Server started with:

```
vllm serve /models/DeepSeek-V4.1-Flash \
  --tensor-parallel-size <TP> --data-parallel-size <DP> --enable-expert-parallel \
  --moe-backend <BACKEND> --tokenizer-mode deepseek_v41 --language-model-only \
  --max-model-len 5120 --max-num-seqs 128 --max-num-batched-tokens 16384 \
  --gpu-memory-utilization 0.95 --kv-cache-dtype fp8 \
  --model-loader-extra-config '{"enable_multithread_load": true, "num_threads": 64}' \
  --no-enable-prefix-caching --no-enable-flashinfer-autotune --seed 42 --port 8000 \
  [--speculative-config '{"method": "dspark", "num_speculative_tokens": 5}']
```

Results after the fix:

| version | TP | DP | BACKEND | speculative-config |
|-----|----|----|---------|--------------------|
| candidate | 1 | 4 | flashinfer_moe_ep_cutedsl | none |
| candidate | 1 | 4 | flashinfer_moe_ep_cutedsl | dspark, 5 tokens |
| candidate | 4 | 1 | flashinfer_moe_ep_cutedsl | dspark, 5 tokens |
| candidate | 1 | 4 | flashinfer_moe_ep_mega_cutedsl | dspark, 5 tokens |

Using the gsm8k eval script provided with vLLM on 4xGB200.

**Comment by jdebache** [comment]
@tlrmchlsmth @bnellnm : thanks for having taken the time to review.

Since the changes now deviate quite a bit from the original changeset, I need to have another look through all of them myself to find potential inconsistencies. I'll drop a ping here again when it's ready for another round if you don't mind.

**Comment by jdebache** [comment]
Hey @bnellnm @tlrmchlsmth: I've had a look through again. I noticed that some of the logic that was in the quantization methods actually belonged to the oracle. I've refactored that.

I will redo the end-to-end testing in the coming days. In the meanwhile, would you mind triggering CI again please?
